### PR TITLE
Dataset creation

### PR DESCRIPTION
**Pull request by jlibovicky**
This PR changes the way the input files are processed while creating a dataset.

* series inputs is a path to a file or a list of file paths
* if the input is different something different than UTF-8 space-tokenized plain text, a reader can be specified in a tuple with the paths
* single series preprocessors are defined by another field which is a list of three-tuples conaining: name of the original series, name of the target series and the function that does the preprocessing

In this way, we can have easily configurable readers for various types of data (plain text, numpy tensors, list of image path, GIZA alignment). Not overwriting the loaded series by the preprocessing, we can better track what is going on inside the preprocessing (e.g., in case of BPE, we will see both the original sentences and those preprocessed by BPE).

Dataset-level preprocessing is still to be done (such that for instance automatic post-editing with edit operations could be trained without generating intermediate files).

After merging this PR, it will be necessary to change all configuration files that are using preprocessing and those using non-textual data.

### CODE REVIEW

File: neuralmonkey/dataset.py [hunk 1]
```diff
@@ -77,28 +91,49 @@ def _get_name_from_paths(series_paths: Dict[str, str]) -> str:
...
-    argument specs.
+def _get_series_paths_and_readers(
+        kwargs: Dict[str, Any]) -> Dict[str, Tuple[List[str], Reader]]:
+    """Get paths to files that contain data from the dataset kwargs.
 
-    Input file for a serie named 'xxx' is specified by parameter 's_xxx'
+    Input file for a serie named 'xxx' is specified by parameter 's_xxx'. The
+    dataset series is define by a string with a path / list of strings with
+    path, or a tuple whose first member is path or list of path and the second
+    memeber is a reader function.
```
Comment (jindrahelcl) [inline neuralmonkey/dataset.py]: Co takhle tam teda přidělat i preprocessory?

Comment (jindrahelcl) [inline neuralmonkey/dataset.py]: Mimochodem, co to teda znamená, že tam můžu dát buď string a to bude cesta, anebo tuple, a první bude cesta nebo seznam cest a druhý bude reader?

File: neuralmonkey/dataset.py [hunk 2]
```diff
@@ -77,28 +91,49 @@ def _get_name_from_paths(series_paths: Dict[str, str]) -> str:
...
-def _get_series_paths(kwargs: Dict[str, str]) -> Dict[str, str]:
-    """Get paths to files that contain data from the dataset keyword
-    argument specs.
+def _get_series_paths_and_readers(
+        kwargs: Dict[str, Any]) -> Dict[str, Tuple[List[str], Reader]]:
+    """Get paths to files that contain data from the dataset kwargs.
 
-    Input file for a serie named 'xxx' is specified by parameter 's_xxx'
+    Input file for a serie named 'xxx' is specified by parameter 's_xxx'. The
+    dataset series is define by a string with a path / list of strings with
```
Comment (jindrahelcl) [inline neuralmonkey/dataset.py]: .. is define**d** ...

File: neuralmonkey/dataset.py [hunk 3]
```diff
@@ -77,28 +91,49 @@ def _get_name_from_paths(series_paths: Dict[str, str]) -> str:
...
+    series_sources = {}
+    for name, key in zip(names, keys):
+        value = kwargs[key]
+
+        if isinstance(value, tuple):
+            paths = value[0]
+            reader = value[1]
+        else:
+            paths = value
+            reader = UtfPlainTextReader
```
Comment (jindrahelcl) [inline neuralmonkey/dataset.py]: todle znamená, že vše, co neni text musí mít explicitně danej reader?

**Comment by jindrahelcl** [comment]
For more heterogenous datasets (e.g. for MMMT), there will need to be an option to add a preprocessor to a subset of series which would operate on all of the series from the subset.

I propose to do it in a way that single-serie preprocessors, multi-serie preprocessors, and dataset-level preproessors are all treated in the same fashion, preferably as these dataset-subset-preprocessors.

File: neuralmonkey/dataset.py (hunk 1 above)
Comment (jlibovicky) [inline neuralmonkey/dataset.py]: Jo.

File: neuralmonkey/dataset.py (hunk 3 above)
Comment (jlibovicky) [inline neuralmonkey/dataset.py]: Jo.

Comment (jindrahelcl) [inline neuralmonkey/dataset.py]: Je.

**Comment by jlibovicky** [comment]
It is a reasonable proposal, indeed, however I would leave it for another PR, so @cifkao can write a reader for GIZA alignment as soon as he wishes.  We can discuss how to preprocessing should ultimately work elswhere (in a next PR).

**Comment by jindrahelcl** [comment]
Ok.

**Comment by jindrahelcl** [comment]
Ne počkej ještě dodělej ty chybičky!!

**Comment by jlibovicky** [comment]
Hotovo.

**Pull request merged**